src/rgw/motr/gc/gc.cc

```diff
@@ -145,33 +150,58 @@ bool MotrGC::going_down() {
   return down_flag;
 }
 
-int MotrGC::dequeue(const DoutPrefixProvider* dpp, std::string iname, motr_gc_obj_info obj)
-{
+int MotrGC::enqueue(motr_gc_obj_info obj) {
+  int rc = 0;
+  // create keys 0_{obj.tag} & 1_{obj.time + min_wait_time}
+  std::string key1 = obj_tag_prefix + obj.tag;
+  std::string key2 = obj_exp_time_prefix +
+                     std::to_string(obj.time + cct->_conf->rgw_gc_obj_min_wait);
+
+  bufferlist bl;
+  obj.encode(bl);
+  // push {0_ObjTag: motr_gc_obj_info} to the gc queue.📥
+  rc = store->do_idx_op_by_name(index_names[enqueue_index],
+                                M0_IC_PUT, key1, bl);
+  if (rc < 0)
+    return rc;
+  
+  // push {1_ExpiryTime: motr_gc_obj_info} to the gc queue.📥
+  rc = store->do_idx_op_by_name(index_names[enqueue_index],
+                                M0_IC_PUT, key2, bl);
+  if (rc < 0)
+    return rc;
+  
+  // rolling increment the enqueue_index
+  enqueue_index = (enqueue_index + 1) % max_indices;
+  return rc;
+}
+
+int MotrGC::dequeue(std::string iname, motr_gc_obj_info obj) {
   int rc;
   bufferlist bl;
-  rc = static_cast<rgw::sal::MotrStore*>(store)->do_idx_op_by_name(iname,
-                                M0_IC_DEL, obj.tag, bl);
+  rc = store->do_idx_op_by_name(iname, M0_IC_DEL, obj.tag, bl);
   if (rc < 0){
     ldout(cct, 0) << "ERROR: failed to delete tag entry "<<obj.tag<<" rc: " << rc << dendl;
   }
-  rc = static_cast<rgw::sal::MotrStore*>(store)->do_idx_op_by_name(iname,
-                                M0_IC_DEL, std::to_string(obj.time), bl);
+  rc = store->do_idx_op_by_name(
+                iname, M0_IC_DEL, std::to_string(obj.time), bl);
   if (rc < 0 && rc != -EEXIST){
     ldout(cct, 0) << "ERROR: failed to delete time entry "<<obj.time<<" rc: " << rc << dendl;
   }
   return rc;
 }
 
-int MotrGC::get_locked_gc_index(uint32_t& rand_ind)
-{
-  int rc = -1;
+int MotrGC::get_locked_gc_index(uint32_t& rand_ind) {
+  int rc = 0;
   uint32_t new_index = 0;
+  bool index_locked = true;
   // attempt to lock GC starting with passed in index
-  for (uint32_t ind = 0; ind < max_indices; ind++)
-  {
+  for (uint32_t ind = 1; ind < max_indices; ind++) {
     new_index = (ind + rand_ind) % max_indices;
     // try locking index
     // on sucess mark rc as 0
+    if (index_locked)
+      break;
```

Comment:
index locked will be initialized as `false` and set to `true` on the successful acquisition of the lock, till then processing any gc queue that is requested.